Core/Src/CAN/CAN_IDs.c

```diff
@@ -1,99 +1,94 @@
-    CAN_ID_map[ID_REGEN_BRAKE_STATE]    = (CAN_ID_t)INTERNAL_ID_REGEN_BRAKE_STATE;
-    CAN_ID_map[ID_SPEED_L_MOTOR]        = (CAN_ID_t)INTERNAL_ID_SPEED_L_MOTOR;
-    CAN_ID_map[ID_SPEED_R_MOTOR]        = (CAN_ID_t)INTERNAL_ID_SPEED_R_MOTOR;
-    CAN_ID_map[ID_POWER_L_MOTOR]        = (CAN_ID_t)INTERNAL_ID_POWER_L_MOTOR;
-    CAN_ID_map[ID_POWER_R_MOTOR]        = (CAN_ID_t)INTERNAL_ID_POWER_R_MOTOR;
-    CAN_ID_map[ID_ENERGY_L_MOTOR]       = (CAN_ID_t)INTERNAL_ID_ENERGY_L_MOTOR;
-    CAN_ID_map[ID_ENERGY_R_MOTOR]       = (CAN_ID_t)INTERNAL_ID_ENERGY_R_MOTOR;
-    CAN_ID_map[ID_OVERLOAD_L_MOTOR]     = (CAN_ID_t)INTERNAL_ID_OVERLOAD_L_MOTOR;
-    CAN_ID_map[ID_OVERLOAD_R_MOTOR]     = (CAN_ID_t)INTERNAL_ID_OVERLOAD_R_MOTOR;
-    CAN_ID_map[ID_LOST_MSG_L_MOTOR]     = (CAN_ID_t)INTERNAL_ID_LOST_MSG_L_MOTOR;
-    CAN_ID_map[ID_LOST_MSG_R_MOTOR]     = (CAN_ID_t)INTERNAL_ID_LOST_MSG_R_MOTOR;
-    CAN_ID_map[ID_BUS_OFF_L_MOTOR]      = (CAN_ID_t)INTERNAL_ID_BUS_OFF_L_MOTOR;
-    CAN_ID_map[ID_BUS_OFF_R_MOTOR]      = (CAN_ID_t)INTERNAL_ID_BUS_OFF_R_MOTOR;
-    CAN_ID_map[ID_CAN_STATE_L_MOTOR]    = (CAN_ID_t)INTERNAL_ID_CAN_STATE_L_MOTOR;
-    CAN_ID_map[ID_CAN_STATE_R_MOTOR]    = (CAN_ID_t)INTERNAL_ID_CAN_STATE_R_MOTOR;
-    CAN_ID_map[ID_INV_STATE_L_MOTOR]    = (CAN_ID_t)INTERNAL_ID_INV_STATE_L_MOTOR;
-    CAN_ID_map[ID_INV_STATE_R_MOTOR]    = (CAN_ID_t)INTERNAL_ID_INV_STATE_R_MOTOR;
-    CAN_ID_map[ID_FAILURE_L_MOTOR]      = (CAN_ID_t)INTERNAL_ID_FAILURE_L_MOTOR;
-    CAN_ID_map[ID_FAILURE_R_MOTOR]      = (CAN_ID_t)INTERNAL_ID_FAILURE_R_MOTOR;
-    CAN_ID_map[ID_ALARM_L_MOTOR]        = (CAN_ID_t)INTERNAL_ID_ALARM_L_MOTOR;
-    CAN_ID_map[ID_ALARM_R_MOTOR]        = (CAN_ID_t)INTERNAL_ID_ALARM_R_MOTOR;
-
-    // define um identificador unico a cada conjunto id & pos. Mapeia esse identificador
-    // em outro vetor, para facil acesso posterior no datalog.
-    for (uint16_t i = 0; i < CAN_ID_QUAN; i++) {
-        CAN_ID_t* current_ID = CAN_ID_map + i;
-        uint16_t unique_index;
-        // ids fora de range, de debug
-        if (current_ID->id > ECU_CAN_LAST_POPULATED_ID) {
-            unique_index = convert_id_pos_to_index(
-                ECU_CAN_LAST_POPULATED_ID + 1 + current_ID->id - ECU_CAN_FIRST_DEBUG_ID,
-                current_ID->pos);
-        } else {
-            unique_index = convert_id_pos_to_index(current_ID->id, current_ID->pos);
+void initialize_CAN_IDs_struct() {
+    initialize_CAN_var_inf();
+    sort_struct();
+    // gets the amount of external id
+    quant_of_ext_id = 1;
+    for (uint16_t i = 0; i < (uint16_t)CAN_GENERAL_ID_QUAN - 1; i++) {
+        if (CAN_ID_map[i].id != CAN_ID_map[i + 1].id) {
+            quant_of_ext_id++;
         }
-        CAN_ID_POS_INTERNAL_ID_MAP[unique_index] = i;
     }
+    populate_datalog_send_struct();
+}
+
+static void populate_datalog_send_struct() {
+    uint16_t j = 0;
+    uint16_t i = 0;
+    for (j = 0; j < quant_of_ext_id; j++) {
+        for (uint16_t k = 0; k < 4; k++) {
```

Comment:
tentar sempre criar define pra numeros
```suggestion
        for (uint16_t k = 0; k < WORDS_PER_ID; k++) {
```